### PR TITLE
Remove error in the state when there is no error anymore.

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are a lot of dependencies, I only list the main class and test code to save reviewers' time. The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big one in the end and merge once after all review PRs get approved. Now the code is missing unit tests.  Will add unit tests, run performance tests, and fix bugs before the official release.

### Description
Previously, we won't reset error in the state when a detector changes from an error state to a non-error state. This PR fixes that.

Testing done:
* Tested on a cluster the problem is gone with the fix.
 
### Check List
- [ Y ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java [hunk 1]
```diff
@@ -468,12 +468,14 @@ private void indexAnomalyResult(
...
+            detectionStateHandler.saveError(response.getError(), detectorId);
             // skipping writing to the result index if not necessary
             // For a single-entity detector, the result is not useful if error is null
             // and rcf score (thus anomaly grade/confidence) is null.
-            // For a multi-entity detector, we don't need to save on the detector level.
-            // We always return 0 rcf score if there is no error.
-            if (response.getAnomalyScore() <= 0 && response.getError() == null) {
+            // For a HCAD detector, we don't need to save on the detector level.
+            // We return 0 or Double.NaN rcf score if there is no error.
+            if ((response.getAnomalyScore() <= 0 || Double.isNaN(response.getAnomalyScore())) && response.getError() == null) {
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java:478]: From the description
> Previously, we won't reset error in the state when a detector changes from an error state to a non-error state. This PR fixes that.

Should we set error as null in this if block?

Comment (kaituo) [reply src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java:478]: You meant call saveError under this branch?

**Review by spbjss — APPROVED** [review]
looks good to me.

File: src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java (hunk 1 above)
Comment (ylwu-amzn) [reply src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java:478]: Oh, I see your force pushed a new commit. Not sure what the original code looks like. But current code looks ok for me.

**Review by ylwu-amzn — APPROVED** [review]
LGTM

**Pull request closed without merging**